File: Java/ZigzagLevelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        boolean left2right = true;

        while(!q.isEmpty()) {
            List<Integer> elem = new ArrayList<>();
            int size = q.size();
            for(int i = 0;i < size; i++) {
                TreeNode n = q.poll();
                elem.add(n.val);
                if(n.left != null) q.offer(n.left);
                if(n.right != null) q.offer(n.right);
            }
            if(!left2right) Collections.reverse(elem);
            res.add(elem);
            left2right = !left2right;
        }

        return res;
    }
}

File: Go/zigzag_level_order.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func reverse(elem []int) []int {
    for i,j := 0, len(elem) - 1; i < j; i,j = i + 1, j - 1 {
        elem[i], elem[j] = elem[j], elem[i]
    }
    return elem
}

func zigzagLevelOrder(root *TreeNode) [][]int {
    if root == nil {
        return nil
    }
    res := [][]int{}
    q := []*TreeNode{}
    q = append(q, root)
    left2right := true

    for len(q) > 0 {
        elem := []int{}
        size := len(q)

        for i := 0; i < size; i++ {
            n := q[i]
            elem = append(elem, n.Val)
            if n.Left != nil {
                q = append(q, n.Left)
            }
            if n.Right != nil {
                q = append(q, n.Right)
            }
        }
        if !left2right {
            elem = reverse(elem)
        }
        res = append(res, elem)
        left2right = !left2right
        q = q[size:]
    }
    return res
}